File: code/python/facade/ControlPanel.py
from AdaptiveCruiseControlSystem import AdaptiveCruiseControlSystem
from AutomaticParkingSystem import AutomaticParkingSystem
from CollisionAvoidanceSystem import CollisionAvoidanceSystem


class ControlPanel:
    def __init__(self):
        self.parkingSystem = AutomaticParkingSystem()
        self.cruiseControlSystem = AdaptiveCruiseControlSystem()
        self.collisionSystem = CollisionAvoidanceSystem()

    def park(self):
        self.parkingSystem.park()

    def activate_cruise_control(self):
        self.cruiseControlSystem.activate()

    def avoid_collision(self):
        self.collisionSystem.avoid()

    def drive_to_home_and_park(self):
        print("Taking you home...")
        self.activate_cruise_control()
        self.avoid_collision()
        self.park()

File: code/typescript/src/facade/ControlPanel.ts
import { AdaptiveCruiseControlSystem } from "./AdaptiveCruiseControlSystem";
import { AutomaticParkingSystem } from "./AutomaticParkingSystem";
import { CollisionAvoidanceSystem } from "./CollisionAvoidanceSystem";

export class ControlPanel {
    private parkingSystem: AutomaticParkingSystem = new AutomaticParkingSystem();
    private cruiseControlSystem: AdaptiveCruiseControlSystem = new AdaptiveCruiseControlSystem();
    private collisionSystem: CollisionAvoidanceSystem = new CollisionAvoidanceSystem();

    park(): void {
        this.parkingSystem.park();
    }

    activateCruiseControl(): void {
        this.cruiseControlSystem.activate();
    }

    avoidCollision(): void {
        this.collisionSystem.avoid();
    }

    driveToHomeAndPark(): void {
        console.log("Taking you home...");
        this.activateCruiseControl();
        this.avoidCollision();
        this.park();
    }
}
